docs/guides/haiku_migration_guide.rst

```diff
@@ -0,0 +1,551 @@
+the model. In Haiku, you use ``hk.transform`` over a function
+that calls your Module, ``transform`` will return an object with ``init``
+and ``apply`` methods. In Flax, you simply instantiate your Module.
+
+.. codediff::
+  :title_left: Haiku
+  :title_right: Flax
+  :sync:
+
+  def forward(x, training: bool):
+    return Model(256, 10)(x, training)
+
+  model = hk.transform(forward)
+
+  ---
+
+  ...
+
+
+  model = Model(256, 10)
+
+To get the model parameters in both libraries you use the ``init`` method
+with a ``PRNGKey`` plus some inputs to run the model. The main difference here is
+that Flax returns a mapping from collection names to nested array dictionaries,
+``params`` is just one of these possible collections. In Haiku, just get the ``params``
+structure directly.
+
+.. codediff::
+  :title_left: Haiku
+  :title_right: Flax
+  :sync:
+
+  sample_x = jax.numpy.ones((1, 784))
+  params = model.init(
+    PRNGKey(0),
+    sample_x, training=False # <== inputs
+  )
+  ...
+
+  ---
+
+  sample_x = jax.numpy.ones((1, 784))
+  variables = model.init(
+    PRNGKey(0),
+    sample_x, training=False # <== inputs
+  )
+  params = variables["params"]
+
+One very important thing to note is that in Flax the parameters structure
+is hierarchical, reflacting the structure of the model. Whereas in Haiku
+the parameters structure is mostly flat, it maps a ``/`` separated string path
+to the parameters of the most inner Modules.
+
+.. tab-set::
+
+  .. tab-item:: Haiku
+    :sync: Haiku
+
+    .. code-block:: python
+
+      ...
+      {
+        'model/block/linear': {
+          'b': (256,),
+          'w': (784, 256),
+        },
+        'model/linear': {
+          'b': (10,),
+          'w': (256, 10),
+        }
+      }
+      ...
+
+
+  .. tab-item:: Flax
+    :sync: Flax
+
+    .. code-block:: python
+
+      FrozenDict({
+        Block_0: {
+          Dense_0: {
+            bias: (256,),
+            kernel: (784, 256),
+          },
+        },
+        Dense_0: {
+          bias: (10,),
+          kernel: (256, 10),
+        },
+      })
+
+During training in both frameworks you pass the parameters structure to the
+``apply`` method to run the forward pass. Since we are using dropout, in
+both cases we must provide a ``key`` to ``apply`` in order to generate
+the random masks.
+
+.. codediff::
+  :title_left: Haiku
+  :title_right: Flax
+  :sync:
+
+  def train_step(key, params, inputs, labels):
+    def loss_fn(params):
+        logits = model.apply(
+          params,
+          key,
+          inputs, training=True # <== inputs
+        )
+        return optax.softmax_cross_entropy_with_integer_labels(logits, labels).mean()
+
+    grads = jax.grad(loss_fn)(params)
+    params = jax.tree_map(lambda p, g: p - 0.1 * g, params, grads)
+
+    return params
+
+  ---
+
+  def train_step(key, params, inputs, labels):
+    def loss_fn(params):
+        logits, updates = model.apply(
```

Comment:
don't want an updates here as no mutables are set

Reply:
thanks for catching this! added a `testcode` block to catch this on CI.